### PR TITLE
Support load rpc parallel & compress

### LINKED ISSUE #3703: load rpc support parallel & compress
### Parallel
Currently StarRocks Loading Coordinator send data write request to BE after last request has response. 
When we don't need to consider the order of loading data, it will become the bottleneck of load. 
So we support parallel data write request when the user needs it. 
set query option `load_parallel_request_num` to enable parallel request, the default value is `1`, the legal value between `1` to `16`.
it also support set `load_parallel_request_num` in stream load HTTP request

### Compress
Currently StarRocks Loading Coordinator send data write request to BE with RAW chunk data,
And users generally use three replicas, it is easy to fill up the network bandwidth in large-scale load scenarios.
So we support compressing write request when user needs it. 
set query option `transmission_compression_type` to enable compress request, the default value is none compress

### PR DESCRIPTION
**Pull request by meegoo**
1. load_parallel_request_num to enable parallel load rpc request
2. transmission_compression_type to enable compression load rpc request

## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] others

## Which issues of this PR fixes ：
Fixes #3703

### Parallel
Currently StarRocks Loading Coordinator send data write request to BE after last request has response.
When we don't need to consider the order of loading data, it will become the bottleneck of load.
So we support parallel data write request when the user needs it.

### Compress
Currently StarRocks Loading Coordinator send data write request to BE with RAW chunk data,
And users generally use three replicas, it is easy to fill up the network bandwidth in large-scale load scenarios.
So we support compressing write request when user needs it.

### CODE REVIEW

File: gensrc/thrift/FrontendService.thrift [hunk 1]
```diff
@@ -523,6 +523,8 @@ struct TStreamLoadPutRequest {
     25: optional i64 thrift_rpc_timeout_ms
     26: optional string json_root
     27: optional bool partial_update
+    28: optional string transmissionCompressionType
```
Comment (imay) [inline gensrc/thrift/FrontendService.thrift:526]: 1. Already have a compression type `enum TCompressionType` define in gensrc/thrift/Types.thrift
2. naming should accord to other fields, better to use `transmission_compression_type`

Comment (meegoo) [reply gensrc/thrift/FrontendService.thrift:526]: do we have any rule of naming, there have both underscore and camel naming in this structure

Comment (imay) [reply gensrc/thrift/FrontendService.thrift:526]: Before if the struct is used by BE this use underscore like cpp, if the struct is used by FE this use camel like Java.

Comment (meegoo) [reply gensrc/thrift/FrontendService.thrift:526]: The problem is that thrift is used by BE and FE both. I will use the underscore to name it uniformly.

File: be/src/exec/tablet_sink.cpp [hunk 2]
```diff
@@ -172,138 +195,226 @@ Status NodeChannel::open_wait() {
...
-                           _pending_batches_num >= _max_pending_batches_num)) {
-        SCOPED_RAW_TIMER(&_mem_exceeded_block_ns);
-        SleepFor(MonoDelta::FromMilliseconds(10));
-    }
+    RETURN_IF_ERROR(_wait_one_prev_request());
 
-    if (_cur_chunk->columns().empty()) {
-        _cur_chunk = chunk->clone_empty_with_slot();
+    if (!eos) {
+        SCOPED_TIMER(_parent->_pack_chunk_timer);
```
Comment (chaoyli) [inline be/src/exec/tablet_sink.cpp:261]: Where is the mem_tracker to tracker the memory consumption?

File: be/src/runtime/tablets_channel.cpp [hunk 3]
```diff
@@ -102,26 +104,15 @@ void TabletsChannel::add_chunk(brpc::Controller* cntl, const PTabletWriterAddChu
...
-        LOG(WARNING) << "Out-of-order request from " << cntl->remote_side() << ". seq=" << request.packet_seq()
-                     << " expect=" << sender.next_seq << " load_id=" << _load_channel->load_id();
-        response->mutable_status()->set_status_code(TStatusCode::INVALID_ARGUMENT);
-        response->mutable_status()->add_error_msgs("out-of-order request");
-        return;
+        if (sender.next_seq < 0) {
+            response->mutable_status()->set_status_code(TStatusCode::INTERNAL_ERROR);
+            response->mutable_status()->add_error_msgs("Tablet channel has been cancelled");
+            return;
+        }
```
Comment (chaoyli) [inline be/src/runtime/tablets_channel.cpp:115]: If the request.packet_seq() is lower than sender.next_seq, what will happen?

**Review by chaoyli — CHANGES_REQUESTED** [review]

File: be/src/runtime/tablets_channel.cpp (hunk 3 above)
Comment (meegoo) [reply be/src/runtime/tablets_channel.cpp:115]: nothing happen

File: be/src/exec/tablet_sink.cpp [hunk 4]
```diff
@@ -172,138 +195,226 @@ Status NodeChannel::open_wait() {
...
+    // 2. check other request unblocking
+    for (size_t i = 0; i < _max_parallel_request_size; i++) {
+        if (_add_batch_closures[i]->count() == 1) {
+            _current_request_index = i;
+            RETURN_IF_ERROR(_wait_request(_add_batch_closures[i]));
+            return Status::OK();
+        }
+    }
+
+    // 3. waiting one request
```
[the remark this answers is not in the document]
Comment (meegoo) [reply be/src/exec/tablet_sink.cpp:391]: step1 & step2 are unblocking

File: be/src/exec/tablet_sink.cpp (hunk 2 above)
Comment (meegoo) [reply be/src/exec/tablet_sink.cpp:261]: automatic tracer by the memory hook, the previous implementation has a individual queue memory tracker but no need now.

File: be/src/exec/tablet_sink.cpp [hunk 5]
```diff
@@ -172,138 +191,276 @@ Status NodeChannel::open_wait() {
...
+    return false;
+}
+
+Status NodeChannel::_wait_one_prev_request() {
+    SCOPED_TIMER(_parent->_wait_response_timer);
+    if (_next_packet_seq == 0) {
+        return Status::OK();
+    }
+
+    // 1. check last request unblocking
```
Comment (chaoyli) [inline be/src/exec/tablet_sink.cpp:423]: 1. check the last request using short-circuit
This may be a better description.

**Review by chaoyli — APPROVED** [review]

**Comment by chaoyli** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 10 / 15 (66.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|5|8|62.50%|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|5|7|71.43%|

File: gensrc/thrift/InternalService.thrift [hunk 6]
```diff
@@ -159,6 +159,8 @@ struct TQueryOptions {
   54: optional i32 pipeline_dop;
   // For pipeline query engine
   55: optional TPipelineProfileLevel pipeline_profile_level;
+  // For load parallel request num
+  56: optional i32 load_parallel_request_num;
```
Comment (chaoyli) [inline gensrc/thrift/InternalService.thrift:163]:
```
load_dop
```
dop is standard for the degree of parallel. The query execution already uses this noun. It's better to be consistent.

**Review by chaoyli — CHANGES_REQUESTED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|7|8|87.50%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|7|8|87.50%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|7|8|87.50%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|7|8|87.50%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 13 / 15 (86.67%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/task/StreamLoadTask.java|6|7|85.71%|
|:large_blue_circle:|com/starrocks/planner/StreamLoadPlanner.java|7|8|87.50%|

**Review by chaoyli — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Pull request merged**